File: mirror/perl5/Perl6.MetaModel/lib/Perl6/Attribute.js

if (Perl6 == undefined) var Perl6 = function () {};

Perl6.Attribute = function (associated_with, label, props) {

    var accessor_name = label.substring(2, label.length);
    var visibility = 'public';
    if (label.indexOf(':') == 1) visibility = 'private';
    if (props == undefined) {
        props = { 'access' : 'ro', 'type' : undefined };
    }
    else {
        if (!props['access']) props['access'] = 'ro';
        if (!props['type'])   props['type']   = undefined;        
    }

    this.associated_with = associated_with;
    this.accessor_name = accessor_name;
    this.visibility = visibility;
    this.properties = props;
    this.label = label;
}

Perl6.Attribute.prototype.is_ro = function () {
    return this.properties['access'] == 'ro';
}

Perl6.Attribute.prototype.is_rw = function () {
    return this.properties['access'] == 'rw';    
}

Perl6.Attribute.prototype.type = function () {
    return this.properties['type'];    
}

Perl6.Attribute.prototype.label = function () {
    return this.label;
}

Perl6.Attribute.prototype.is_array = function () {
    return this.label.charAt(0) == '@';
}

Perl6.Attribute.prototype.is_hash = function () {
    return this.label.charAt(0) == '%';    
}

Perl6.Attribute.prototype.associated_with = function () {
    return this.assocaited_with;
}

Perl6.Attribute.prototype.accessor_name = function () {
    return this.accessor_name;    
}

Perl6.Attribute.prototype.is_private = function () {
    return this.visibility == 'private';        
}

Perl6.Attribute.prototype.is_public = function () {
    return this.visibility == 'public';            
}

Perl6.Attribute.prototype.instantiate_container = function () {
    if (this.is_array()) return [];
    if (this.is_hash())  return {};    
    return undefined;
}

/*


*/

File: mirror/perl5/org.perl6.metamodel/src/org/perl6/metamodel/Attribute.java

package org.perl6.metamodel;

import java.util.*;

public class Attribute {
    
    private MetaClass associated_with;
    private String accessor_name;
    private String visibility;
    private HashMap properties;
    private String label;
    
    public Attribute (MetaClass _associated_with, String _label)
    {
        this(_associated_with, _label, new HashMap());
    }
    
    public Attribute (MetaClass _associated_with, 
                      String _label, 
                      HashMap _properties)
    {
        String _accessor_name = _label.substring(2, _label.length());
        String _visibility = "public";
        
        if (_label.indexOf(':') == 1)
            _visibility = "private";

        if(!_properties.containsKey("access"))
            _properties.put("access","ro");
        if(!_properties.containsKey("type"))
            _properties.put("type", null);
        
        this.associated_with = _associated_with;
        this.accessor_name = _accessor_name;
        this.visibility = _visibility;
        this.label = _label;
        this.properties = _properties;
    }
    
    public boolean is_ro () {
        return ( (String) properties.get("access") ).equals("ro");
    }
    
    public boolean is_rw () {
        return ( (String) properties.get("access") ).equals("rw");
    }
    
    public String type () {
        return (String) properties.get("type");
    }
    
    public String label () {
        return label;
    }
    
    public boolean is_array () {
        return label.charAt(0) == '@';
    }
    
    public boolean is_hash () {
        return label.charAt(0) == '%';
    }
            
    public MetaClass associated_with () {
        return associated_with;
    }
    
    public String accessor_name () {
        return accessor_name;
    }
    
    public boolean is_private () {
        return visibility.equals("private");
    }
    
    public boolean is_public () {
        return visibility.equals("public");
    }
    
    public java.lang.Object instantiate_container () {
        if (this.is_array()) {
            return new ArrayList();
        }
        if (this.is_hash()) {
            return new HashMap();
        }
        return null;
    }
}